oak-lucene/src/test/java/org/apache/jackrabbit/oak/plugins/index/lucene/dynamicBoost/FulltextQueryTermsProviderImpl.java

```diff
@@ -0,0 +1,150 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.jackrabbit.oak.plugins.index.lucene.dynamicBoost;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import org.apache.jackrabbit.oak.plugins.index.lucene.spi.FulltextQueryTermsProvider;
+import org.apache.jackrabbit.oak.spi.state.NodeState;
+import org.apache.lucene.analysis.Analyzer;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.TermQuery;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An example fulltext query terms provider.
+ */
+public class FulltextQueryTermsProviderImpl implements FulltextQueryTermsProvider {
+    
+    private static final Logger LOG = LoggerFactory.getLogger(FulltextQueryTermsProviderImpl.class);
+    private static final String SEARCH_SPLIT_REGEX = "[ ]";
+    private static final String NT_DAM_ASSET = "dam:Asset";
+    private static final String PREDICTED_TAGS_REL_PATH = "";
+    private static final int MAX_FRAGMENT_SIZE = 2;
+    private static final int MAX_QUERY_SIZE = 10;
+    
+    @Override
+    public Set<String> getSupportedTypes() {
+        Set<String> supportedTypes = new HashSet<String>();
+        supportedTypes.add(NT_DAM_ASSET);
+        return supportedTypes;
+    }
+
+    @Override
+    public org.apache.lucene.search.Query getQueryTerm(String text, Analyzer analyzer, NodeState indexDefinition) {
+        if (analyzer == null || text == null) {
+            return null;
+        }
+
+        LOG.debug("getQueryTerm Text: {}", text);
+        BooleanQuery query = this.createQuery();
+
+        Set<String> charTerms = new HashSet<String>(splitForSearch(text));
+        LOG.debug("getQueryTerm charTerms: {}", charTerms);
+        if(charTerms.size() > MAX_QUERY_SIZE) {
+            LOG.debug("Not adding query terms for smart tags as number of terms in the query {} exceeds " +
+                    "maximum permissible value of {}", charTerms.size(), MAX_QUERY_SIZE);
+            return null;
+        }
+        List<String> fragments = prepareFragments(charTerms);
+
+        for(String fragment : fragments) {
+            Term term = new Term(PREDICTED_TAGS_REL_PATH + fragment.toLowerCase(), "1");
+            query.add(new TermQuery(term), BooleanClause.Occur.SHOULD);
+            LOG.debug("Added query term: {}", fragment.toLowerCase());
+        }
+
+
+        Term term = new Term(PREDICTED_TAGS_REL_PATH + text.toLowerCase(), "1");
+        query.add(new TermQuery(term), BooleanClause.Occur.SHOULD);
+        LOG.debug("Added query term: {}", text.toLowerCase());
+
+        //De-boosting smart tags based query. CQ-4194419
+        query.setBoost(0.0001f);
+        return query;
```

Comment:
The class isn't actually used currently, I could remove it from the patch. It will be needed once we implement the "query" part of this feature in Oak.

But I wouldn't currently want to change how it's implemented... It is what is in use _currently_. We want to replace it with something else, for sure; but we should keep it in the test source folder so we can compare the new and the old implementation. We can change the behavior slightly in the Oak feature, but not here.